HW#5 - Nikita Chkhaidze

nchkhaidze opened the pull request:
Made a basic graphical interface to simplify user inputs.

Link below for testing:

https://jsfiddle.net/njahnik/jxuh5fgy/7/

nvkuznetsova commented on script.js:
```diff
@@ -0,0 +1,140 @@
+
+Order.prototype.pay = function () {
+    this.paid = true;
+    this.editable = false;
```
You can use `Object.freeze` to make an order not editable

nvkuznetsova commented on script.js:
```diff
@@ -0,0 +1,140 @@
+    this.size = size;
+    this.stuffing = stuffing;
+    if (this.size == 'large') {
+        FoodItem.call(this, 'Large Hamburger with ' + this.stuffing.name, 100 + this.stuffing.price, 40 + this.stuffing.calories);
```
You can use methods to calculate the price and calories.

nvkuznetsova commented:
So, your code works but that is not exactly what should have been done. You need to create separate classes for Hamburger, Salad, and Drink. You can create types for all types of menu items and stuffing like in task description. All these classes are inherited from the base class FoodItem. Also in your solution price and calories of Salad items calculated incorrectly as Salad should have property `weight` as mentioned in the description.

Try to rewrite your code according to this UML (not the best example but I hope you get the main idea. Properties and methods with the same names as in the base class are inherited. And you can override the inherited methods to extend its functionality)
![image](https://user-images.githubusercontent.com/31770506/98726811-0a58c600-23a8-11eb-964c-581bc779d32d.png)

Also, please look through this article https://studfile.net/preview/6161982/page:59/. This is a synopsis of an article from an old learn.javascript tutorial (it's really hard to find the original article after the tutorial has been updated). Good luck!😊

nchkhaidze commented:
@nvkuznetsova 

Thank you very much for the feedback. 

Commited corrected version of the task. Added classes for Drink, Salad and Stuffing items. Added getPrice and getCalories methods which are special for Hamburger and Salad class. Added Object.freeze for pay() method.

However, I found that when I use object.create for setting the prototype as in: 

`Salad.prototype = Object.create(FoodItem.prototype);
Salad.prototype.constructor = Salad;`

It didn't work afterwards with instanceof operator which I used for Order.prototype.addItem and  Order.prototype.showItems.
So i kept the setting of prototype through __proto__ as in the previous version.

link in JSFiddle: https://jsfiddle.net/njahnik/jxuh5fgy/10/

nvkuznetsova commented:
Your code looks much better now! 😊 
I suggested a solution to the problem with the `instanceof` operator.
But something when wrong when I tried to add salad using UI. I added 1 Olivie salad and got this output
![image](https://user-images.githubusercontent.com/31770506/99192924-b161a700-2786-11eb-842f-eeb766555c41.png)
But the correct output is
![image](https://user-images.githubusercontent.com/31770506/99192932-be7e9600-2786-11eb-986e-c25d833f07d6.png)
However, for 1 Cesar salad, I got correct output
![image](https://user-images.githubusercontent.com/31770506/99192956-dce49180-2786-11eb-99bd-18484747afad.png)
I see you add weight in `addItem` of the `Order` class. Maybe you shouldn't initialize weight with 100 in the `Salad` class constructor
![image](https://user-images.githubusercontent.com/31770506/99193006-246b1d80-2787-11eb-876a-42e60663b764.png)
Try to initialize weight with 0. And please add `setWeight` method to `Salad` class. It's better to modify this property using a method in `addItem` method.

nchkhaidze commented:
@nvkuznetsova 

Thank you very much for remarks.
I fixed the bug with salad prices, replaced __proto__ with Object.create and implemented getType and setWeight methods.
New link in JsFiddle:
https://jsfiddle.net/njahnik/jxuh5fgy/17/
